### ISSUE TITLE
[Bugs] Elusive sessions error

### ISSUE THREAD

**Issue by mystralkk** [Geeklog-Core/geeklog#121]
**Reported by mst3kroqs on 12 May 2008 23:09**

**Description:**
OK - I didn't imagine it this time. 

As you know- I've been bashing on 1.5.0b1 for you for the past several days, and have experienced an oddity where (and I use FireFox for almost all bugbashing b/c FireBug is so darn useful for finding css/script/dom stuff) occasionally I would experience the following:
1. login as admin. 
2. work with the site for awhile.
3. logout.
4. login as a different user
5. work with the site for awhile.
6. logout.
7. login as admin

After logging in as admin the second time, the site renders content as if I were an anonymous user, although the cookies seem set to the admin values (eg. gl_password cookie hash is appropriate for admin as per db), and 'Admin' shows up under logged in user, and there also seems to be an entry in the gl_sessions table with a sess_id equal to the gl_session cookie.

Oh yeah - btw - I rename my cookies with gl_ prefixes just to be sure they don't conflict with other '3rd party' apps that may be installed on the site that also utilize separate auth mechanisms.

I tried retrieving an admin-only URL just to see if it was a rendering problem, but the site really has not assigned me admin rights. Wierd.

After this, I cannot login as any other user either, all equate to anonymous, however the session info goes into gl_sessions. The only way I seem to be able to clear the error is to remove all of the cookies for the test site.

NOTE: This is not immediately reproduceable. Some period of time has to pass before the bug manifests itself.

Also - I have classified this as a crash because today, I caught it in the act of crashing the SQL server, and wonder if this somehow related. The server restarted itself, and all seemed well, but this was in /logs/error.log:

Mon May 12 18:51:46 2008 - 1053: Server shutdown in progress. SQL in question: DELETE FROM gl_sessions WHERE start_time < 1210632406

BTW- the DELETE seems to have succeeded - the gl_sessions table was empty except for the one (current) session.

Wierd, huh. Random event?

[Mantis Bugtracker #630](http://project.geeklog.net/tracking/view.php?id=630)

**Comment by mystralkk** [comment]
**Comment posted by mst3kroqs on 13 May 2008 15:28**

I have just reproduced the problem.  In this case, I was logged in as admin.  I logged out, and logged back in with my named user account, to which I have assigned the &#039;root&#039; privilege.

After doing this, I can see the session entry in gl_sessions, and my account name shows up in the &#039;Who&#039;s Online&#039; block, but I get no User Functions or Admins Only blocks on the left, and I do not have admin privileges.

This is a &#039;stock&#039; 1.5.0b1 install, the only &#039;mod&#039; is the installation of the caching template library, as I was testing Nouveau compatibility at the time.

There was no SQL server crash this time, but I&#039;m going to try to continue to exercise the site to see if I can reproduce that.

**Comment by mystralkk** [comment]
**Comment posted by dhaun on 13 May 2008 15:41**

Sounds like the presence of a _caching_ template library would explain the effects you are seeing quite nicely ...

Can you reproduce the problem without it?

**Comment by mystralkk** [comment]
**Comment posted by mst3kroqs on 13 May 2008 18:04**

Hi Dirk -  nominally, I would agree, and I have already started installing a &#039;clean&#039; 1.5.0b1 testsite in anticipation of your request, however consider that:

1) I had disabled the CTL the last time the error happened, eg. the old template class and professional theme was being used.  The CTL template code was there, but not in use.

2) Even after logging in, I was not assigned the appropriate privileges to receive the User Functions or Admins Only blocks.  I was also unable to manually retrieve the moderation page.

3) I don&#039;t understand how the presence of the CTL code would lead to the SQL server crash I documented.

In other news, there have been multiple reports of session table difficulties in the past (not sure if it is related) but consider:

http://www.geeklog.net/forum/viewtopic.php?forum=3&amp;showtopic=78765
http://www.geeklog.net/forum/viewtopic.php?forum=3&amp;showtopic=77272
http://www.geeklog.net/forum/viewtopic.php?forum=3&amp;showtopic=69347
http://www.gllabs.org/forum/viewtopic.php?showtopic=8435

Clearly there is something going on here, and it isn&#039;t good.  ;^)

Anyway - I will do the clean install and try to reproduce again.

**Comment by mystralkk** [comment]
**Comment posted by jmucchiello on 13 May 2008 21:07**

Dirk, the CTL does not cache data. It caches the template. It turns

{site_url}

into

&lt;?php echo $this-&gt;get_var(&#039;site_url&#039;); ?&gt;

How could that affect your login?

**Comment by mystralkk** [comment]
**Comment posted by jmucchiello on 13 May 2008 21:09**

There are other gl_session issues in the wild:

http://www.gllabs.org/forum/viewtopic.php?showtopic=8435

**Comment by mystralkk** [comment]
**Comment posted by mst3kroqs on 18 May 2008 01:35**

Note that I am not discounting that lib-sessions could use an overhaul, but I believe that the root cause for this sessions-related error was the use of mod_fcgid in conjunction with testing the site on my hosting provider.  

I will send you a more complete explanation by e-Mail, but it appears that although mod_fcgid is supposed to be &#039;script-transparent&#039;, it may be corrupting DOM data (eg. $_SESSIONS[], $_GET[], $_POST[].

I utilized this mode when testing sites, as the repeated pageloads caused by my webapp security scan tools would often generate CPU exceeded errors.

Note that I also believe this is related to the &#039;false positive&#039; I got when testing kses.class.php, because the page I was using was repeatedly reloaded in order to try all test vectors.

Anyway - this was quite an adventure, finding this one, but overall I think that GL might want to consider giving folks a word of warning relative to using this feature.  In general, don&#039;t use it, but some hosting providers may not provide a choice, and errors associated with it may be attributed to GL.

The easiest way to find out if you are running fastcgi, is to invoke phpinfo(), and look at the &#039;Configure Command&#039; string, if you see &#039;--enable-fastcgi&#039;, then it&#039;s probably running.

There may be a way to disable it by utilizing a modified .htaccess, users should contact their hosting provider.

Hope this helps!

For info on mod_fcgid, see: http://fastcgi.coremail.cn/

**Comment by mystralkk** [comment]
**Comment posted by dhaun on 18 May 2008 09:32**

There&#039;s no doubt that the session code is ancient, inefficient, and not very useful for anything other than keeping the user logged in. On the other hand, it has gone through a few security fixes and is well tested and trusted in that regard. Any changes in the session code are likely to open it up to security issues again, which is one reason for the lack of motivation to touch it ...

All the linked cases of corruption of the session table seem to follow the same pattern: MySQL crashed and when it came back up, the session table was corrupted (or possibly got corrupted on first access). Frankly, I fail to see how that could be Geeklog&#039;s fault. SQL requests could be inefficient or wrong, but they should never be able to make the DB server crash. If it does crash, that is a problem in the DB server, IMO.

Incidentally, we did have the occassional gl_sessions corruption on geeklog.net in the past, but I haven&#039;t seen one since the last time the server was updated. Which makes me think that it only happens with certain versions of MySQL and again seems to point at a possible bug in the DB server.

The bit about mod_fcgid was new to me, though. It might be worth starting a little survey amongst those that suffer from gl_sessions corruption about the MySQL version they&#039;re on and whether or not they are using mod_fcgid.
